sdk/core/Azure.Core/src/Internal/ArrayBackedPropertyBag.cs

```diff
@@ -0,0 +1,140 @@
+                    }
+                    _key2 = key;
+                    _value2 = value;
+                    break;
+
+                default:
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key1!))
+                    {
+                        duplicateKeyDetected = true;
+                        _value1 = value;
+                        break;
+                    }
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key2!))
+                    {
+                        duplicateKeyDetected = true;
+                        _value2 = value;
+                        break;
+                    }
+                    _array ??= new (TKey Key, TValue Value)[8];
+                    if (_arrayCount >= _array.Length)
+                    {
+                        // The array must be re-sized
+                        (TKey Key, TValue Value)[] newItems = new (TKey Key, TValue Value)[_array.Length * 2];
+                        Array.Copy(_array, newItems, _array.Length);
+                        _array = newItems;
+                    }
+                    _array[_arrayCount] = new(key, value);
+                    _arrayCount++;
+                    break;
+            }
+            if (!duplicateKeyDetected)
+            {
+                _totalCount++;
+            }
+        }
+
+        /// <summary>
+        /// Gets a value from the property bag.
+        /// </summary>
+        /// <param name="key">The key of the item to get.</param>
+        /// <param name="value">The out parameter that will store the value, if found.</param>
+        /// <returns><c>true</c> if found, else <c>false</c>.</returns>
+        public bool TryGetValue(TKey key, out TValue? value)
+        {
+            Argument.AssertNotNull(key, nameof(key));
+
+            switch (_totalCount)
+            {
+                case 0:
+                    value = default;
+                    return false;
+
+                case 1:
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key1!))
+                    {
+                        value = _value1;
+                        return true;
+                    }
+
+                    value = default;
+                    return false;
+
+                default:
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key1!))
+                    {
+                        value = _value1;
+                        return true;
+                    }
+
+                    if (EqualityComparer<TKey>.Default.Equals(key, _key2!))
+                    {
+                        value = _value2;
+                        return true;
+                    }
+
+                    if (_array == null)
+                    {
+                        value = default;
+                        return false;
+                    }
+
+                    // search the array in reverse to favor the last item added in the case of duplicate keys.
```

Comment:
OH, then why do you even track if keys are duplicated?

Reply:
for the case where the dupe is in the first two items (not in the array)

Reply:
I'm not sure I understand why we optimize for writing rather than for reading. Is it a common case that we store values for the keys that are never used?